File: backend/app/schemas/changelog.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class ChangeHistoryField(BaseModel):
    """Field change in a changelog entry."""

    name: str
    previous: Any = None
    current: Any = None


class LatestChange(BaseModel):
    """Latest change details for a vulnerability."""

    changed_at: str = Field(serialization_alias="changedAt")
    change_type: str = Field(serialization_alias="changeType")
    job_name: str = Field(serialization_alias="jobName")
    job_label: str | None = Field(default=None, serialization_alias="jobLabel")
    fields: list[ChangeHistoryField] = Field(default_factory=list)

    model_config = {"populate_by_name": True}


class ChangelogEntry(BaseModel):
    """
    Represents a single changelog entry for a vulnerability.
    """

    vuln_id: str = Field(serialization_alias="vulnId")
    title: str
    source: str
    change_type: str = Field(
        serialization_alias="changeType",
        description="Type of change: 'created' or 'updated'",
    )
    timestamp: datetime
    cvss_score: float | None = Field(default=None, serialization_alias="cvssScore")
    severity: str | None = None
    latest_change: LatestChange | None = Field(
        default=None, serialization_alias="latestChange"
    )

    model_config = {"populate_by_name": True}


class ChangelogResponse(BaseModel):
    """
    Response containing a list of recent vulnerability changes.
    """

    entries: list[ChangelogEntry]
    total: int

File: frontend/src/api/changelog.ts
import { api } from "./client";

export interface ChangeHistoryField {
  name: string;
  previous: any;
  current: any;
}

export interface LatestChange {
  changedAt: string;
  changeType: string;
  jobName: string;
  jobLabel: string | null;
  fields: ChangeHistoryField[];
}

export interface ChangelogEntry {
  vulnId: string;
  title: string;
  source: string;
  changeType: "created" | "updated";
  timestamp: string;
  cvssScore?: number;
  severity?: string;
  latestChange?: LatestChange;
}

export interface ChangelogResponse {
  entries: ChangelogEntry[];
  total: number;
}

export const fetchChangelog = async (
  limit: number = 50,
  offset: number = 0,
  fromDate?: string,
  toDate?: string,
  source?: string,
): Promise<ChangelogResponse> => {
  const params: Record<string, string | number> = { limit, offset };
  if (fromDate) params.fromDate = fromDate;
  if (toDate) params.toDate = toDate;
  if (source) params.source = source;
  const response = await api.get<ChangelogResponse>("/v1/changelog", { params });
  return response.data;
};
